### PR TITLE
Support hidden fields in code generation

### PR DESCRIPTION
**Pull request by ArsenHD**
# Description

This PR fixes issue #646. Hidden fields have been supported in the engine by @volivan239 . This PR adds support for them on the code generation side.

Let's consider an example. We have `class A { int x; }` and `class B extends A { int x; }`. These fields `x` have the same name, but they are __different__ fields. Let's say we have a variable (of type `A` or `B`) and we want to access field `x` (from class `A` or `B`). There are 4 cases:
- Variable of type `A` accessing `x` of type `A`. Here field can be accessed directly.
- Variable of type `A` accessing `x` of type `B`. Here we need a cast to type `B` and then we are able to access the field.
- Variable of type `B` accessing `x` of type `A`. Here we also need a cast, but to type `A`.
- Variable of type `B` accessing `x` of type `B`. Here field can be accessed directly.

This logic is now implemented in `CgCallableAccessManagerImpl`. This class checks if we can call a method with the given arguments and if we can access a field using the given object. Previously, there were no checks for fields, only for methods. Now we check field for accessibility. There are 3 cases:
- Field can be accessed directly (`myObj.field`)
- Field can be accessed with type cast (`((MyClass) myObj).field`)
- Field can only be accessed via reflection (e.g. private fields)

## Type of Change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

## Automated Testing

These tests check the hidden fields functionality.

`org.utbot.examples.objects.HiddenFieldAccessModifiersTest`
`org.utbot.examples.objects.HiddenFieldExampleTest`

## Manual Scenario 

TODO

# Checklist (remove irrelevant options):

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered
- [x] No new warnings
- [ ] New tests have been added
- [ ] All tests pass locally with my changes

### CODE REVIEW

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt [hunk 1]
```diff
@@ -120,6 +127,26 @@ internal class CgCallableAccessManagerImpl(val context: CgContext) : CgCallableA
...
+                    fieldId
+                )
+            }
+            // we can access the field only via reflection
+            is ReflectionOnly -> fieldId.accessWithReflection(this)
+        }
+    }
+
+    override operator fun ClassId.get(fieldId: FieldId): CgStaticFieldAccess {
+        return CgStaticFieldAccess(fieldId)
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt:147]: Minor: Maybe use expression body here?

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt [hunk 2]
```diff
@@ -182,13 +209,15 @@ internal class CgCallableAccessManagerImpl(val context: CgContext) : CgCallableA
...
     }
 
-    private infix fun CgExpression?.canBeReceiverOf(executable: MethodId): Boolean =
-        when {
-            // TODO: rewrite by using CgMethodId, etc.
-            outerMostTestClass == executable.classId && this isThisInstanceOf outerMostTestClass -> true
-            executable.isStatic -> true
-            else -> this?.type?.isSubtypeOf(executable.classId) ?: false
+    private infix fun CgExpression.canBeReceiverOf(executable: MethodId): Boolean {
+        return when {
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt:213]: Minor: Maybe use expression body here?

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt [hunk 3]
```diff
@@ -264,10 +293,115 @@ internal class CgCallableAccessManagerImpl(val context: CgContext) : CgCallableA
...
+        }
+
+        // check arguments suitability
+        if (!(args canBeArgsOf this)) {
+            return false
+        }
+
+        // If method is static, then it must not have a caller.
+        if (this.isStatic) {
+            require(caller == null) { "Caller expression of a static method call must be null" }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt:309]: Caller is not always null - we can invoke static methods with instance in Java

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt [hunk 4]
```diff
@@ -264,10 +293,115 @@ internal class CgCallableAccessManagerImpl(val context: CgContext) : CgCallableA
...
+        }
+
+        requireNotNull(caller) { "Method must have a caller, unless it is the method of the current test class or a static method" }
+        return caller canBeReceiverOf this
+    }
+
+    private fun FieldId.accessSuitability(accessor: CgExpression?): FieldAccessorSuitability {
+        // if field is static
+        if (this.isStatic) {
+            require(accessor == null) { "Accessor expression of a static field access must be null" }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt:325]: Accessor is not always null - we can access static fields with instance in Java

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt [hunk 5]
```diff
@@ -264,10 +293,115 @@ internal class CgCallableAccessManagerImpl(val context: CgContext) : CgCallableA
...
+            return Suitable
+        }
+
+        val fieldDeclaringClassType = this.declaringClass
+        val accessorType = accessor.type
+
+        if (fieldDeclaringClassType is BuiltinClassId || accessorType is BuiltinClassId) {
+            // The rest of the logic of this method processes hidden fields.
+            // We cannot check the fields of builtin classes, so we assume that we work correctly with them,
+            // because they are well known library classes (e.g. Mockito) that are unlikely to have hidden fields.
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt:359]: Could you move this comment after 361 line, please?

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt [hunk 6]
```diff
@@ -264,10 +293,115 @@ internal class CgCallableAccessManagerImpl(val context: CgContext) : CgCallableA
...
+            return Suitable
+        }
+
+        val fieldName = this.name
+
+        if (accessorType isSubtypeOf fieldDeclaringClassType || fieldDeclaringClassType isSubtypeOf accessorType) {
+            if (fieldName.isNameOfFieldsInBothClasses(accessorType, fieldDeclaringClassType)) {
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt:366]: Looks like it does not work - for example, field "x" in private inheritor of some class "A" would be declared with `Object` type, and we will return `Suitable` here which is wrong

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt [hunk 7]
```diff
@@ -352,7 +351,7 @@ internal class CgVariableConstructor(val context: CgContext) :
         val init = if (classId.isAccessibleFrom(testClassPackageName)) {
             CgGetJavaClass(classId)
         } else {
-            classId[forName](classId.name)
+            Class::class.id[forName](classId.name)
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt:354]: `classClassId`

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt [hunk 8]
```diff
@@ -284,6 +305,46 @@ internal class CgStatementConstructorImpl(context: CgContext) :
         currentBlock += buildCgForEachLoop(init)
     }
 
+    override fun createFieldVariable(fieldId: FieldId): CgVariable {
+        val declaringClass = newVar(Class::class.id) { Class::class.id[forName](fieldId.declaringClass.name) }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt:309]: `classClassId`

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt [hunk 9]
```diff
@@ -284,6 +305,46 @@ internal class CgStatementConstructorImpl(context: CgContext) :
...
+        val name = fieldId.name + "Field"
+        return newVar(java.lang.reflect.Field::class.id, name) {
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt:311]: `fieldClassId`?

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt [hunk 10]
```diff
@@ -284,6 +305,46 @@ internal class CgStatementConstructorImpl(context: CgContext) :
...
+            declaringClass[getDeclaredField](fieldId.name)
+        }
+    }
+
+    override fun createExecutableVariable(executableId: ExecutableId, arguments: List<CgExpression>): CgVariable {
+        val declaringClass = newVar(Class::class.id) { Class::class.id[forName](executableId.classId.name) }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt:317]: `classClassId`

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt [hunk 11]
```diff
@@ -284,6 +305,46 @@ internal class CgStatementConstructorImpl(context: CgContext) :
...
+        val argTypes = (arguments zip executableId.parameters).map { (argument, parameterType) ->
+            val baseName = when (argument) {
+                is CgVariable -> "${argument.name}Type"
+                else -> "${parameterType.prettifiedName.decapitalize()}Type"
+            }
+            newVar(classCgClassId, baseName) {
+                if (parameterType.isPrimitive) {
+                    CgGetJavaClass(parameterType)
+                } else {
+                    Class::class.id[forName](parameterType.name)
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt:327]: `classClassId`

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt [hunk 12]
```diff
@@ -284,6 +305,46 @@ internal class CgStatementConstructorImpl(context: CgContext) :
...
+                }
+            }
+        }
+
+        return when (executableId) {
+            is MethodId -> {
+                val name = executableId.name + "Method"
+                newVar(java.lang.reflect.Method::class.id, name) {
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt:335]: `methodClassId`?

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt [hunk 13]
```diff
@@ -284,6 +305,46 @@ internal class CgStatementConstructorImpl(context: CgContext) :
...
+                    declaringClass[getDeclaredMethod](executableId.name, *argTypes.toTypedArray())
+                }
+            }
+            is ConstructorId -> {
+                val name = executableId.classId.prettifiedName.decapitalize() + "Constructor"
+                newVar(java.lang.reflect.Constructor::class.id, name) {
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/util/CgStatementConstructor.kt:341]: `constructorClassId`?

**Review by Damtev — CHANGES_REQUESTED** [review]
A problem with RequiresTypeCast?

**Review by Damtev — CHANGES_REQUESTED** [review]

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt [hunk 14]
```diff
@@ -224,10 +254,179 @@ internal class CgCallableAccessManagerImpl(val context: CgContext) : CgCallableA
...
+     * `A` and `I1` do not have any relationship with each other, but `B` **can** access fields `x` from both of them.
+     * However, it **must** be done with a type cast (to `A` or `I1`), because otherwise accessing field `x` will
+     * be ambiguous. Method [isFieldHiddenIn] **does consider** such cases as well.
+     *
+     * **These examples show** that when checking if a field is hidden we **must** consider all supertypes
+     * in all "branches" **up to** the type where the field is declared (**exclusively**).
+     * Then we check if any of these types has a field with the same name.
+     * If such field is found, then the field is hidden, otherwise - not.
+     */
+    private infix fun FieldId.isFieldHiddenIn(subclass: Class<*>): Boolean {
```
[the remark this answers is not in the document]
Comment (ArsenHD) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgCallableAccessManager.kt:419]: [Done](https://github.com/UnitTestBot/UTBotJava/pull/763/commits/caba5d10a165006873c0c7734a1878038a9910d1)

**Review by Damtev — APPROVED** [review]

**Pull request merged**